### PR TITLE
Initialize large table value randomly

### LINKED ISSUE #9777: uniform_random_op supports selected_rows

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #9777 
issue: https://github.com/PaddlePaddle/Paddle/issues/9211

### CODE REVIEW

File: paddle/fluid/operators/uniform_random_table_op.cc [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    auto shard_id = Attr<int>("shard_id");
+    auto max_id = Attr<int>("max_id");
+    auto shape = Attr<std::vector<int>>("shape");
+
+    auto tensor = out->mutable_value();
+    tensor->Resize(framework::make_ddim(shape));
+    // Only allocate the memory of large table on CPU
+    auto cpu = platform::CPUPlace();
+    float *data = tensor->mutable_data<float>(cpu);
+    VLOG(3) << "generate seed";
```
Comment (typhoonzero) [inline paddle/fluid/operators/uniform_random_table_op.cc]: VLOGs not needed.

File: paddle/fluid/operators/uniform_random_table_op.cc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        scope.FindVar(Output("Out"))->GetMutable<framework::SelectedRows>();
+    auto shard_cnt = Attr<int>("shard_cnt");
+    auto shard_id = Attr<int>("shard_id");
+    auto max_id = Attr<int>("max_id");
+    auto shape = Attr<std::vector<int>>("shape");
+
+    auto tensor = out->mutable_value();
+    tensor->Resize(framework::make_ddim(shape));
+    // Only allocate the memory of large table on CPU
+    auto cpu = platform::CPUPlace();
```
Comment (typhoonzero) [inline paddle/fluid/operators/uniform_random_table_op.cc]: Do we need to enforce that `dev_place` is on CPU?

File: paddle/fluid/operators/uniform_random_table_op.cc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    } else {
+      rows_size = max_id / shard_cnt + 1;
+    }
+    auto *rows = out->mutable_rows();
+    rows->resize(rows_size);
+    (*rows)[0] = shard_id;
+    for (int64_t idx = 1; idx < rows_size; ++idx) {
+      (*rows)[idx] = (*rows)[idx - 1] + shard_cnt;
+    }
+    out->set_height(max_id);
```
Comment (typhoonzero) [inline paddle/fluid/operators/uniform_random_table_op.cc]: As discussed offline, with @Yancey1989 @jacquesqiao , need to implement this using an auto-growth `SelectedRows`, and init it randomly with a startup init buffer size, like 128M. This feature may also need to implement operations that when "Prefetch" runs.

File: paddle/fluid/operators/uniform_random_table_op.cc [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    auto shard_cnt = Attr<int>("shard_cnt");
+    auto shard_id = Attr<int>("shard_id");
+    auto max_id = Attr<int>("max_id");
+    auto shape = Attr<std::vector<int>>("shape");
+
+    auto tensor = out->mutable_value();
+    tensor->Resize(framework::make_ddim(shape));
+    // Only allocate the memory of large table on CPU
+    auto cpu = platform::CPUPlace();
+    float *data = tensor->mutable_data<float>(cpu);
```
Comment (jacquesqiao) [inline paddle/fluid/operators/uniform_random_table_op.cc]: This `float` should be a template

File: paddle/fluid/operators/uniform_random_table_op.cc [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    VLOG(3) << "RunImpl...";
+    auto out =
+        scope.FindVar(Output("Out"))->GetMutable<framework::SelectedRows>();
+    auto shard_cnt = Attr<int>("shard_cnt");
+    auto shard_id = Attr<int>("shard_id");
+    auto max_id = Attr<int>("max_id");
+    auto shape = Attr<std::vector<int>>("shape");
+
+    auto tensor = out->mutable_value();
+    tensor->Resize(framework::make_ddim(shape));
```
Comment (jacquesqiao) [inline paddle/fluid/operators/uniform_random_table_op.cc]: The first dimension of the tensor is not sure, should be calculated by some attribute, such as id_num / shard_cnt + buffer_size. Do you mean that this calculation is done in Python ?

File: paddle/fluid/operators/uniform_random_table_op.cc (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/uniform_random_table_op.cc]: when lookup table op find that some id is not in the rows of the table, it should push this new id to the table.rows to init it.

Comment (Yancey1989) [reply paddle/fluid/operators/uniform_random_table_op.cc]: FROM @typhoonzero 
> and init it randomly with a startup init buffer size, like 128M

We have already had an attribute `shape` to represent the shape of the `SelectedRows.value()`, maybe we can only use the `shape` to determinate the memory size, use `shape` and `memory size` at the same time may be confused.

Comment (Yancey1989) [reply paddle/fluid/operators/uniform_random_table_op.cc]: And maybe we resue `uniform_random_op`, they have the same computing logic.

**Comment by Yancey1989** [comment]
@jacquesqiao @typhoonzero 
I updated this PR and allocate an initialize buffer by the attribute `shape`, and then we need to refine `lookup_table_op` to re-allocate the Table's memory if new Id was coming and not enough memory.

I created an issue #9841 and will implement it ASAP.

File: paddle/fluid/operators/uniform_random_op.cc [hunk 6]
```diff
@@ -24,7 +24,17 @@ template <typename T>
...
+    framework::Tensor* tensor(nullptr);
+    auto out_var = ctx.OutputVar("Out");
+    if (out_var->IsType<framework::LoDTensor>()) {
+      tensor = out_var->GetMutable<framework::LoDTensor>();
+    } else if (out_var->IsType<framework::SelectedRows>()) {
+      auto shape = ctx.Attr<std::vector<int>>("shape");
+      tensor = out_var->GetMutable<framework::SelectedRows>()->mutable_value();
+      tensor->Resize(framework::make_ddim(shape));
+    } else {
+      PADDLE_THROW("Only support SelectedRows and Tensor");
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply paddle/fluid/operators/uniform_random_op.cc]: Done.

**Pull request merged**